### DISCUSSION TITLE
Design patterns for benchmarking with FIO

### DISCUSSION THREAD

**Discussion by vsoch** [axboe/fio#1635] [Q&A] [answered]
This question is followup from [this tweet](https://twitter.com/axboe/status/1703493404800565293?s=46&t=p1ZYYP24NpcQBMP4kUSfVw)

I'm new to benchmarking I/O and am enjoying using FIO. I was wondering if folks have good references that describe differences between concurrent vs. random vs. sequential, but in the context of FIO? E.g., it would be cool to see the different design patterns, and especially in some graphical form.

I am definitely finding a lot of generic resources (e.g., [here is an example](https://educative.io/answers/difference-between-sequential-and-random-access-to-storage) but I don't want to misrepresent what FIO is doing in case there is variance from that. 

Thank you!

**Comment by axboe** [comment, marked as answer]
In terms of access patterns, sequential IO should be pretty self explanatory - it's IO where the start of the next IO lines up with the end of the previous one. Eg you read/write 4k at offset 0, next IO will start at offset 4k.

For random, there are some options and differences. By default, fio will randomize over the entire IO range, using a selectable randomness generator (tausworth (default) or lfsr), which can be set with ```random_generator```. For each IO, a new offset is picked at random. Fio will try and hit the entire range, maintaining state of which blocks have been hit. This is controllable with ```norandommap```.

Since purely random IO isn't necessarily that real world relevant outside of micro benchmarks, fio also supports different distributions of randomness. This is selectable with ```random_distribution``` and defaults to ```random```, which gives you the fully random behavior outlined above. You can also have it mimic distributions like pareto of zipf, or even a normal distribution.
